analyzers/src/SonarAnalyzer.CSharp/Rules/InfiniteRecursion_RoslynCfg.cs

```diff
@@ -0,0 +1,224 @@
+                        context.SemanticModel,
+                        () => context.AnalysisContext.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, context.IssueLocation, "method's recursion")))
+                {
+                    existReferenceToDeclaringSymbolMatcher = new BooleanBlockMatcher(operation =>
+                    {
+                        if (!(operation.Instance.Syntax is InvocationExpressionSyntax invocation))
+                        {
+                            return false;
+                        }
+
+                        return IsInstructionOnThisAndMatchesDeclaringSymbol(invocation.Expression, declaringSymbol, semanticModel);
+                    });
+                }
+
+                protected override bool BlockHasReferenceToDeclaringSymbol(BasicBlock block) =>
+                    existReferenceToDeclaringSymbolMatcher.Matches(block);
+            }
+
+            private class CfgWalkerForProperty : CfgRecursionSearcher
+            {
+                private readonly bool isSet;
+                private readonly BooleanBlockMatcher existReferenceToDeclaringSymbolMatcher;
+
+                public CfgWalkerForProperty(RecursionAnalysisContext<ControlFlowGraph> context, string reportOn, bool isSetAccessor)
+                    : base(
+                        context.ControlFlowGraph,
+                        context.AnalyzedSymbol,
+                        context.SemanticModel,
+                        () => context.AnalysisContext.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, context.IssueLocation, reportOn)))
+                {
+                    isSet = isSetAccessor;
+                    existReferenceToDeclaringSymbolMatcher = new BooleanBlockMatcher(operation =>
+                        operation.Instance.Syntax is var syntax
+                        && TypesForReference.Contains(syntax.GetType())
+                        && MatchesAccessor(syntax)
+                        && IsInstructionOnThisAndMatchesDeclaringSymbol(syntax, declaringSymbol, semanticModel));
+                }
+
+                private static readonly ISet<Type> TypesForReference = new HashSet<Type> { typeof(IdentifierNameSyntax), typeof(MemberAccessExpressionSyntax) };
+
+                protected override bool BlockHasReferenceToDeclaringSymbol(BasicBlock block) =>
+                    existReferenceToDeclaringSymbolMatcher.Matches(block);
+
+                private bool MatchesAccessor(SyntaxNode node)
+                {
+                    var expr = (ExpressionSyntax)node;
+                    var propertyAccess = expr.GetSelfOrTopParenthesizedExpression();
+                    var isNodeASet = propertyAccess.Parent is AssignmentExpressionSyntax assignment && assignment.Left == propertyAccess;
+                    return isNodeASet == isSet;
+                }
+            }
+
+            private abstract class CfgRecursionSearcher : CfgAllPathValidator
+            {
```

Comment:
Both inheritors declare `existReferenceToDeclaringSymbolMatcher` and override `BlockHasReferenceToDeclaringSymbol` in exactly the same way. Let's make `Func<..> referenceToDeclaringSymbolMatcher` a constructor argument that will build a private field `referenceToDeclaringSymbolFinder` here.

Reply:
This was done just halfway. I wanted `RecursionSearcher` to have constructor like this. That should remove some duplications.
```C#
protected RecursionSearcher(RecursionContext<ControlFlowGraph> context, Func<IOperationWrapperSonar, bool> referenceToDeclaringSymbolMatcher)
{
   // ...
   referenceToDeclaringSymbolFinder = new BooleanBlockFinder(referenceToDeclaringSymbolMatcher);
}
```

Reply:
??? There is no BooleanBlockFinder now. This was removed while switching to operations

Reply:
I was reviewing per commit :)